### PR TITLE
Chap. 3: Evaluation, Resampling and Benchmarking (v3)

### PR DESCRIPTION
**Pull request by jemus42**
Next iteration after review on 2023-02-28.

Focus points: Introduction, quick intro (rename to "Holdout and Scoring"

### CODE REVIEW

**Comment by sebffischer** [comment]
I am just reading the chapter and there are some comments from me: 

1. I don't think we should write "This section of the book might be complex for some readers. " but rephrase this to something like "This section covers advanced material". 
2. I think when it comes to the grouping, an example would be nice. Instead of mentioning abstract things like "we want to keep certain observations together", "spatial" etc. I think it would be much nicer to come up with a concrete example and then illustrate using a real-world example why we need grouping. E.g. a medical example where we have repeated observations of individuals but want to make predictions for new patients or something.

**Comment by jemus42** [comment]
Sure, but that's a global thing via the `_complex.qmd` include: https://github.com/mlr-org/mlr3book/blob/main/book/_complex.qmd and affects all chapters, so I'd rather tweak the text of those boxes in the main branch.

**Comment by RaphaelS1** [comment]
> Sure, but that's a global thing via the `_complex.qmd` include: https://github.com/mlr-org/mlr3book/blob/main/book/_complex.qmd and affects all chapters, so I'd rather tweak the text of those boxes in the main branch.

This has already been changed in another PR

**Comment by jemus42** [comment]
Merged main one last time, should be good to go now @RaphaelS1

File: book/performance.qmd [hunk 1]
```diff
@@ -447,12 +483,12 @@ custom_cv
...
 
 
-### Resampling with Stratification and Grouping
+### Resampling with Stratification and Grouping {#sec-strat-group}
 
 {{< include _optional.qmd >}}
 
-In `r mlr3`, we can assign a special role to a feature contained in the data by configuring the corresponding `$col_roles` field of a `r ref("Task")`.
-The two relevant column roles that will affect behavior of a resampling strategy are `"group"` or `"stratum"`.
+In `r mlr3`, we can assign a special role to a feature contained in the data by configuring the corresponding `$col_roles` field of a `r ref("Task")`, which we introduced in @sec-col-roles.
```
Comment (RaphaelS1) [inline book/performance.qmd:490]: Broken reference, sec-col-roles

File: book/performance.qmd [hunk 2]
```diff
@@ -492,14 +528,8 @@ table(task_grp$data(rows = r$test_set(2), cols = "year"))
 table(task_grp$data(rows = r$test_set(3), cols = "year"))
```
Comment (RaphaelS1) [inline book/performance.qmd:528]: Don't need to show all 3, one is enough

File: book/performance.qmd [hunk 3]
```diff
@@ -855,27 +865,84 @@ autoplot(bma, type = "cd")
...
 <!-- Binary classification is unique because of the presence of a positive and negative class and a threshold probability to distinguish between the two. -->
-`r index("ROC")` (Receiver Operating Characteristic) analysis is widely used to evaluate binary classifiers.
-Although extensions for multiclass classifiers exist (see e.g., @hand2001simple), we will only cover the much easier binary classification case here.
-For binary classifiers that predict discrete classes, we can compute a `r index("confusion matrix")` which computes the following quantities (see also @fig-confusion):
 
-* **True positives (TP)**: Instances that are actually positive and correctly classified as positive.
-* **True negatives (TN)**: Instances that are actually negative and correctly classified as negative.
-* **False positives (FP)**: Instances that are actually negative but incorrectly classified as positive.
-* **False negatives (FN)**: Instances that are actually positive but incorrectly classified as negative.
+For the special case of binary classification, there are specialized performance measures and visualization tools to analyze and compare the performance of classifiers.
```
Comment (RaphaelS1) [inline book/performance.qmd:876]: There's specialised performance measures for all tasks. I think could be better just to say something like "We will now look at some specialized performance measures for binary classification that are implemented in mlr3"

**Review by RaphaelS1 — COMMENTED** [review]
* [x] Exercise 2 is too hard also this reads strangely "and they assigned rows using the task's `row_id` modulo 3 ".  I suggest rephrasing to say something like "They said they used a custom 3-fold cross-validation with `factor(task$row_ids %% 3)`. See if you can reproduce their results." -- Currently your questions feels like a 'trick' question as the harder part is basic R and not mlr3
* [ ] You're missing a very basic exercise just to use `$score` - please make this exercise 1
* [ ] Please cleanup code output for solutions


Mainly notes for myself when I takeover:

* Content is very good and structure is great however more verbose than needed. I see a lot of places to cutdown words and repeated code.
* Need to update fit-alts everywhere
* Need a lot more subheadings for paragraphs to help users find text and need to use `define`, etc. and other placeholders
* Tempted to remove 'optional' callout compeltely

**Comment by jemus42** [comment]
> You're missing a very basic exercise just to use $score - please make this exercise 1

It looks like chapter 2 already has an exercise where `$score()` is used. I'm happy to add a simpler ex. 1 here but I'd rather double check what we want it to be before I add something that's too redundant?
I thought about maybe using `mtcars` with a simple `regr.lm` and... I'm not sure, holdout? 3-fold CV? Both, one after the other?

>  Please cleanup code output for solutions
Could you clarify what's messy? I've suppressed warnings (due to `glm.fit`) in one chunk now and reformatted the code slightly but I'm not sure what else to do.

File: book/performance.qmd [hunk 4]
```diff
@@ -34,144 +35,184 @@ set.seed(3)
...
-<!-- Resampling strategies such as $k$-fold cross-validation are usually preferable to a single train-test split because they use all available data to assess the performance and thereby avoid the performance estimate from being biased by a particular split. -->
-<!-- These strategies evaluate the performance of the learning algorithm that induced the final model, rather than the performance of the final model itself. -->
-<!-- However, the best we can do if we only have access to a limited amount of data is to use the performance of the learning algorithm as a proxy for the performance of the final model. -->
+In @sec-basics, we have already applied the holdout method by manually partitioning the data contained in a `r ref("Task")` object into a single training set (to train the model) and a single test set (to estimate the generalization performance).
+Below, we again summarize the entire performance estimation process of the holdout method, from partitioning the task and choosing a performance measure to calculating the score to estimate the generalization performance.
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:76]: In our reviews with Bernd he expressed a preference in favor of some redundancy with chap 2 - the idea being that chap 2 gives a broader intro and here in chap we start from almost-scratch and build up in more detail.

Will recheck with Bernd but just wanted to make sure we're on the same page here.

File: book/performance.qmd [hunk 5]
```diff
@@ -34,144 +35,184 @@ set.seed(3)
...
-Similarly, some measures require the learner to predict probabilities, while others require class predictions.
-In our learner above, we have already selected `predict_type = "prob"`, which is often required for measures that are not defined on class labels, such as the aforementioned AUC.
-
 ::: {.callout-tip}
-More information about a performance measure, including its mathematical definition, can be obtained using the `$help()` method of a `r ref("Measure")` object, which opens the help page of the corresponding measure, e.g., `msr("classif.acc")$help()` provides all information about the classification accuracy.
+We will typically look for a measure that fits our `task_type` at hand (e.g., `"classif"` for the `penguins` task).
+Looking at the `task_properties` columns is useful to identify performance measures that support only binary classification tasks (indicated by `"twoclass"` in the `task_properties` column) such as the `"classif.auc"` measure (see also @sec-roc-space for more details on the AUC).
+For multiclass classification tasks, generalizations of the AUC measure are also available but need to be selected explicitly (e.g., `"classif.mauc_au1p"`).
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:109]: I've now made this part the callout tip and the parts before and after are regular text - that might need to be reformatted a little, but I don't have a strong preference in that regard.

File: book/performance.qmd [hunk 6]
```diff
@@ -34,144 +35,184 @@ set.seed(3)
...
+```
+
+It is also possible to create a list of multiple measures using the `msrs()` function and pass it to the `$score()` method to compare a vector of scores. 
+A `r ref("Measure")` object can also provide other information such as the time it took to train the model or the time to make predictions (e.g., using the measures with the `key`s `time_train` or `time_predict`).
+These measures do not use of the information contained in the `r ref("Prediction")` object and therefore require passing the trained learner to the `$score()` method, which includes the corresponding elapsed time information:
+
+```{r}
+measures = msrs(c("time_train", "time_predict"))
+pred$score(measures, learner = learner)
+```
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:155]: Tried to split this up a little, but didn't want to compress the second part too much.

File: book/performance.qmd [hunk 7]
```diff
@@ -202,92 +244,91 @@ For example, the column `params` shows the parameters of each resampling strateg
...
+```{r}
+rsmp("bootstrap", repeats = 100, ratio = 0.9)
+```
 
-<!-- In our example we're using the quite fast `rpart` learner and the `"penguins"` task with less than 400 observations, where cross-validation should not be an issue. -->
-<!-- If we intended to evaluate a large neural network on an image classification task however, we might not have the computational budget to re-train a learner repeatedly. -->
+ <!-- The `$is_instantiated` field of a `r ref("Resampling")` object constructed as shown above is set to `FALSE`.
+This means that the resampling strategy is not yet applied to a task, i.e., the train-test splits are not contained in the `r ref("Resampling")` object. -->
 
 ### Instantiation {#sec-resampling-inst}
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:293]: Does the first paragraph not suffice?
> "To actually generate the train-test splits for a given task, we need to instantiate a resampling strategy"

I thought it was clear that instantiation is required if you're doing it "manually", but both `resample()` and `benchmark()` take care of that for you. 
Then again, I've been re-reading these sections for months and might have gone numb 😅 
I've tried to add something but it feels a little awkward.

File: book/performance.qmd [hunk 8]
```diff
@@ -350,32 +389,31 @@ While we are usually interested in this aggregated score, it can be useful to lo
...
+We can access the respective `prediction` column or directly use the `$predictions()` method of the `r ref("ResampleResult")` object (without converting it to a `r ref("data.table")` first) to obtain a list of `r ref("Prediction")` objects of each resampling iteration.
+For example, the code below shows the first list element and shows that we get the same result using the `$predictions()` method:
+
+<!-- TODO: Reduce vertical size / avoid figure taking up entire page -->
 
 ```{r performance-019}
 #| eval: true
-rrdt$prediction
-all.equal(rrdt$prediction, rr$predictions())
+rrdt$prediction[[1]]
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:421]: I tend to agree. I've commented that part out for now.

File: book/performance.qmd (hunk 2 above)
Comment (jemus42) [reply book/performance.qmd:538]: Reduced to only show the first test set.

**Comment by berndbischl** [comment]
I just went thru all comment from you, @RaphaelS1, while speaking with @jemus42 on zoom. lukas has no "biggie" question which must be discussed with me and i am fine with all of @RaphaelS1 's comments -- except for 2 open questions.

1) "tempted to remove 'optional' callout completely" --> 
don't get what that means, this sounds like a "global" issue if it refers to "skippable chaps". if so, lets pls resolve / decide this quickly

2) I don't understand why we're doing this again if we've spent all of Chapter 2 covering it? ---> expected a problem here :)
this was a VERY lengthy discussion, also with you editors in chat. i will re-read the section critically now with lukas to see if we can cut it down further

**Comment by berndbischl** [comment]
I have discussed it with Lukas, @RaphaelS1 we think you are mainly / partially correct.

* I think a bit (!!!!!) of summary and redundancy is needed for the reader. 
* chap2-funds was missing some details on the score fun and the measure dict -- and you guys where heavily in favor of putting this in chap3-eval -- which we now did. 

my proposal is this 

* we tell the reader not "we now summarize the complete train-eval-score" workflow but say: "hey we summarize it very BRIEFLY, and we do this BECAUSE we want dive a bit deeper into it to add some missing details on score" 
(explain to the reader WHY we are doing stuff)

text like this:
"We then create aLearnerobject (here a decision tree from therpartpackage) and train theintermediate model using only the training set. After training, we can use the$predictmethod oftheLearnerobject to make predictions based on the test set and store the resultingPredictionobject which will be later used to calculate the performance measure:"

is removed.  the relating code blocks(s) (train predict) are "merged". no intermediate results.

we then continue in the text with
"The next step is to select a suitable performance measure" 

most of what comes here is more detailed then in chap2 and i see value in this. 
@jemus42 can try to cut a few sentences of "fat" there, but i don't think this is that much possible / so worthwhile.

this should shorten the 4-page section by approx 1-1.5 pages. 

d'accord with you @RaphaelS1 ?

**Comment by berndbischl** [comment]
@RaphaelS1 lukas and me have tried to cut out as much "redundant fat" as possible, he will show you a new version asap

**Comment by berndbischl** [comment]
i would be then "gone" here now -- if there's no other or new stuff

File: book/performance.qmd [hunk 9]
```diff
@@ -34,144 +35,184 @@ set.seed(3)
...
+#| fig-align: "center"
+#| fig-alt: "A general abstraction of the performance estimation process: The available data is (repeatedly) split into (a set of) training data and test data (data splitting / resampling process). The learner is applied to each training data and produces intermediate models (learning process). Each intermediate model along with its associated test data produces predictions. The performance measure compares these predictions with the associated actual target values from each test data and computes a performance value for each test data. All performance values are aggregated into a scalar value to estimate the generalization performance (evaluation process)."
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:192]: Ah yes, I wrote alt-text for e.g. `fig-score-aggregate-resampling` and realized how hard it is to do - I tried to fix it here but happy to workshop it more.

**Comment by RaphaelS1** [comment]
> d'accord with you @RaphaelS1 ?

Yeah sounds good!

File: book/performance.qmd [hunk 10]
```diff
@@ -309,7 +350,6 @@ If we do not explicitly pass a `r ref("Measure")` object to the `$score()` metho
 :::
 
 Similarly, we can pass `r ref("Measure")` objects to the `$aggregate()` method to calculate an aggregated score across all resampling iterations.
-The type of aggregation is usually determined by the `r ref("Measure")` object (see also the fields `$average` and `$aggregator` the in help page of `r ref("Measure")` for more details).
 There are two approaches for aggregating scores across resampling iterations: The first is referred to as the `r index("macro average")`, which first calculates the measure in each resampling iteration separately, and then averages these scores across all iterations.
```
[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:347]: I keep staring at that section and fiddleing around with some small code examples, but I can't quite figure out how to show the difference compactly. Using mlr3 objects and doing it manually seems unwieldly and verbose, and manually constucting a dummy set of predictions seems out of place (also also needlessly verbose?). Not sure how to improve that so far :/

[the remark this answers is not in the document]
Comment (RaphaelS1) [reply book/performance.qmd:347]: How about the below, which I *think* is an analogy to micro/macros

```
mean(c(1:3, 4:6))
mean(mean(1:3), mean(4:6))
```

[the remark this answers is not in the document]
Comment (jemus42) [reply book/performance.qmd:347]: I guess you're right -- I've been overcomplicating this way too much in my head, this should suffice as an analogy

**Comment by jemus42** [comment]
So, regarding the ex 1. comment (see also https://github.com/mlr-org/mlr3book/pull/538#issuecomment-1505525935).
Maybe something like this?
My idea was
a) Something simpel (as @RaphaelS1 suggested) primarily using `$score()`
b) Not just re-iterate what @larskotthoff has done in chap 2 exercises
c) Use `resample` in a common and "simple" case (compared to other exercises using `benchmark()` and custom resampling, respectively)


> 1. Apply the `"bootstrap"` resampling strategy on the `mtcars` task and evaluate the performance of the `classif.rpart` decision tree learner.
Use 100 replicates and an a sampling ratio of 80%.
Calculate the MSE for each iteration and visualize the result.
Finally, calculate the aggregated performance score.

```{r}
set.seed(3)
task = tsk("mtcars")
learner = lrn("regr.rpart")
resampling = rsmp("bootstrap", repeats = 100, ratio = 0.8)

rr = resample(task, learner, resampling)
rr$score(msr("regr.mse"))
autoplot(rr)
# Alternatively: Histogram
autoplot(rr, type = "histogram")

rr$aggregate(msr("regr.mse"))
```

Aside from this exercise point I think I am finished with addressing review comments (see also https://github.com/mlr-org/mlr3book/issues/440#issuecomment-1507691750) if I haven't missed anything major.

**Comment by larskotthoff** [comment]
- [ ] 3.2 mention that in LOOCV the number of folds is equal to the number of observations. This will also clarify the "N models" (the text doesn't say what N is).
- [ ] 3.2.2 show an example of how to change `iters` as it's not a param.
- [ ] 3.3.2 Warning: 'as.BenchmarkAggr' is deprecated.
- [ ] 3.3.3 example at the beginning removes "uhash" from output but then references it in the text.
- [ ] Figure 3.7 typography of "axis" labels looks bad.
- [ ] 3.4.3 threshold plots text and plots disagree (changing threshold from 0.5 to 0.75 does make a difference).

**Comment by jemus42** [comment]
> 3.2.2 show an example of how to change iters as it's not a param.

I'm a little confused by that as I a) Did not know you could change `iters` after the fact b) Realized you _can't_ change `iters` after the fact (kind of):

``` r
library(mlr3)

holdout = rsmp("holdout")
holdout$iters <- 30
holdout
#> <ResamplingHoldout>: Holdout
#> * Iterations: 30
#> * Instantiated: FALSE
#> * Parameters: ratio=0.6667
```

So now we've created a subsampling rsmp the long way around? Do we want to show readers that's (pointlessly) possible?
Because the inverse doesn't work:

```r
subsampling = rsmp("subsampling")
subsampling$iters <- 1
#> Error in assert_ro_binding(rhs): Field/Binding is read-only
```

So you can start with `holdout` and "turn it into" subsampling, but you can't take subsampling at turn it into holdout the same way (you could though by changing the `repeats` param of course).
I'm not sure we're doing the readers any favors by pointing out that _sometimes_ it's possible to change `iters`, if it's not a `param_set` value... which happens to have a different name.


- 3.3.3 example at the beginning removes "uhash" from output but then references it in the text:
    - I didn't want to remove it from the text so I explicitly printed it to show it, but I'm not sure if it would be nicer to just hide and not mention it at all.
- Figure 3.7 typography of "axis" labels looks bad
    - I agree, I asked in the book-illustration channel about maybe iterating on this. the latex source is available, should be possible to make it look more in line with the rest of the book.

**Comment by larskotthoff** [comment]
If `iters` shouldn't be changed, then just mention that it's informational and can't be changed :)

I think we can ignore that you can technically change it.

**Comment by jemus42** [comment]
Okay, I tried to clarify that somewhat.

As far as text comments are concerned I'd be done then, which only leaves the issue of Figure 3.7.
Assuming the other changes are considered ok, I'd guess the chapter could move on to the next stage, as an update to the figure should likely not be a blocking issue?

**Review by larskotthoff — APPROVED** [review]
